index.js

```diff
@@ -1,76 +1,119 @@
+const authBuffer = Buffer.from(usernameAndPassword, 'utf-8');
+const requests = [];
+let createdLinksCount = 0;
+let preexistingLinksCount = 0;
+let errCount = 0;
 
-// Send HTTP requests to Jira
 const processRow = row => {
 
-    // Q: What about versions? Do we need to update on all versions?
+  const pageId = row["Page ID"];
+  const title = row["Page Title"];
+  const metadata = row["text_val"];
 
-    const pageId = row["Page ID"];
-    const title = row["Page Title"];
-    const metadata = row["text_val"];
+  const issueKeys = metadata.match(ISSUE_KEY_REGEX);
+  log(INFO, `[PAGE-ID: ${pageId}] Keys: ${issueKeys}`);
 
-    const issueKeys = metadata.match(ISSUE_KEY_REGEX);
-    console.log("Keys: ", issueKeys);
+  const body = {
+    globalId: `appId=${appId}&pageId=${pageId}`,
+    application: {
+      type: "com.atlassian.confluence",
+      name: "Confluence"
+    },
+    relationship: "mentioned in",
+    object: {
+      url: `${confluenceEndpoint}/pages/viewpage.action?pageId=${pageId}`,
+      title
+    }
+  };
 
-    const body = {
-        globalId: `appId=${appId}&pageId=${pageId}`,
-        application: {
-            type: "com.atlassian.confluence",
-            name: "Confluence"
-        },
-        relationship: "mentioned in",
-        object: {
-            url: `${confluenceEndpoint}/pages/viewpage.action?pageId=${pageId}`,
-            title
-        }
-    };
+  issueKeys.forEach(issueKey => {
+    const requestUrl = `${jiraEndpoint}/rest/api/latest/issue/${issueKey}/remotelink`;
 
-    console.log(body);
+    if (isDryRun) {
+      log(INFO, `[DRY-RUN] Will update ${requestUrl}`);
+      return;
+    }
 
-    issueKeys.forEach(issueKey => {
-        const requestUrl = `${jiraEndpoint}/rest/api/latest/issue/${issueKey}/remotelink`;
-        console.log(requestUrl);
-    });
+    requests.push(
+      axios.post(requestUrl, body, {
+        headers: {
+          'Authorization': `Basic ${authBuffer.toString('base64')}`,
+          'Content-Type': 'application/json'
+        }
+      })
+        .then(result => {
+          result.status === 201 ? createdLinksCount++ : preexistingLinksCount++;
+          log(INFO, `Update success: ${requestUrl} => ${result.status} ${JSON.stringify(result.data)}`);
+        })
+        .catch(error => {
+          errCount++
+          log(ERROR, JSON.stringify(body.object));
+          if (error.response) {
+            log(ERROR, `[ERROR] ${requestUrl} => ${error.response.status}`);
+          }
+        })
+    )
+  });
 };
 
 fs.createReadStream(path.resolve(__dirname, csvFilename))
-    .pipe(csv.parse({ headers: true }))
-    .on("error", error => console.error(error))
-    .on("data", processRow)
-    .on("end", rowCount => console.log(`Parsed ${rowCount} rows`));
+  .pipe(csv.parse({headers: true}))
+  .on("error", console.error)
+  .on("data", processRow)
+  .on("end", async rowCount => {
+    log(INFO, `Parsed ${rowCount} rows`);
+    await Promise.allSettled(requests);
```

Comment:
When using `allSettled`, you'll get a list of `resolved` and `rejected` promises so technically you can just count the number of rejected promises and not have to keep your own `errCount`.

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled#using_promise.allsettled.

Not blocking the PR approval

Reply:
This suggestion would only be possible if I were saving promises from axios to the array and then using `Promise.allSettled` to resolve them at one go. However, In this case, each axios request has its own callbacks and `.catch` block. Because of this local catching, all promises return `fulfilled`. Therefore, the `rejected` count is always 0.

I initially wrote the success and failure callbacks here based on the `fulfilled` or `rejected` status but then moved it back locally to each request to be able to log more context about the process being executed. As an example, the `rejected` reason looks like this making it impossible to know which request of the possibly hundreds this originated from
![Screenshot 2021-03-26 at 12 56 50](https://user-images.githubusercontent.com/44353593/112584593-19aa4e80-8e33-11eb-96ce-7ec294c2b1db.jpg)